src/utils/polygon.cpp

```diff
@@ -101,25 +101,33 @@ void Polygons::makeConvex()
         {
             continue; //Already convex.
         }
-        Polygon convexified;
-
-        //Start from a vertex that is known to be on the convex hull: The one with the lowest X.
-        const size_t start_index = std::min_element(poly.begin(), poly.end(), [](Point a, Point b) { return a.X == b.X ? a.Y < b.Y : a.X < b.X; }) - poly.begin();
-        convexified.path->push_back(poly[start_index]);
 
-        for(size_t i = 1; i <= poly.size(); ++ i)
-        {
-            const Point& current = poly[(start_index + i) % poly.size()];
+        auto makeSortedPolyConvex = [](PolygonRef& poly, Polygon& convexified) {
+            convexified.path->push_back(poly[0]);
 
-            //Track backwards to make sure we haven't been in a concave pocket for multiple vertices already.
-            while(convexified.size() >= 2 && LinearAlg2D::pointIsLeftOfLine(convexified.path->back(), (*convexified.path)[convexified.size() - 2], current) >= 0)
+            for(size_t i = 0; i < poly.size() - 1; i ++)
             {
-                convexified.path->pop_back();
+                const Point& current = poly[i];
+                const Point& after = poly[i + 1];
+                if(LinearAlg2D::pointIsLeftOfLine(current, convexified.path->back(), after) < 0)
+                {
+                    //Track backwards to make sure we haven't been in a concave pocket for multiple vertices already.
+                    while(convexified.size() >= 2 && LinearAlg2D::pointIsLeftOfLine(convexified.path->back(), (*convexified.path)[convexified.size() - 2], current) > 0)
+                    {
+                        convexified.path->pop_back();
+                    }
+                    convexified.path->push_back(current);
+                }
             }
-            convexified.path->push_back(current);
-        }
-        //remove last vertex as the starting vertex is added in the last iteration of the loop
-        convexified.path->pop_back();
+        };
+
+        Polygon convexified;
+
+        std::sort (poly.begin(), poly.end(), [](Point a, Point b) { return a.X == b.X ? a.Y < b.Y : a.X < b.X; });
```

Comment:
Knowing both the [Monotone Chain](https://en.wikibooks.org/wiki/Algorithm_Implementation/Geometry/Convex_hull/Monotone_chain) and the [Graham Scan for simple polygons](https://en.wikipedia.org/wiki/Convex_hull_of_a_simple_polygon#Algorithms), why is it necessary here to sort the whole set and not just find the leftmost vertex? I.e. why use Monotone Chain instead of the Graham Scan?

It seems that finding the leftmost vertex here should already be sufficient. And if not, the Monotone Chain shouldn't need to backtrack.

Reply:
If the previous algorithm was supposed to be Graham's scan it was implemented incorrectly. In Grahams scan you're supposed to find the vertex with lowest x-coord. Lets call this point `p_0`. Then after all remaining points should be sorted on "polar angle" relative to `p_0`.

This last step went wrong, we didn't sort but we traversed the points starting from `p_0` in the same order as the original polygon. This order of the points isn't guaranteed to be in order of polar angle.

So we either sort the vertices on polar angle before traversing the vertices, or implement the monotone chain algorithm. In my opinion the monotone chain is easier to understand/implement.

| Monotone Chain shouldn't need to backtrack.
You mean that we should only add points that are "right turning", and its not needed to traverse backwards to remove "left turning" vertices? Yes we do, here is a counter example to that claim

![Untitled Diagram drawio](https://user-images.githubusercontent.com/6638028/172633215-14c2158d-cf4a-47af-a320-e357b0d337a1.png)

vertices `<p_0,p_1,p_2>, <p_1,p_2,p_3>, <p_2,p_3,p_4>` are all right turning and would thus initially not be rejected. Only after backtracking from `p_5` we see that `p_1 ... p_4` should be rejected.